lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,536 @@
+    def create(**kwargs):
+        """
+        create A simple factory method to construct a new Dataset object and
+        add it to the database. Errors are hidden here, and we return None.
+
+        TODO: Ideally, we'd hook up to a logger here and log errors.
+
+        Returns:
+            Dataset: A new Dataset object initialized with the keyword
+            parameters. (See `__init__` for details.)
+        """
+        try:
+            dataset = Dataset(**kwargs)
+            dataset.add()
+        except Exception:
+            return None
+        else:
+            return dataset
+
+    @staticmethod
+    def attach(**kwargs):
+        """
+        attach Attempt to fetch dataset for the controller and dataset name,
+        or using a specified file path (see __init__ and _render_path).
+
+        If the `state` kwarg is specified, the dataset will be created in that
+        state or, if it already exists, `attach` will attempt to advance the
+        dataset to that state.
+
+        Raises:
+            DatasetSqlError: problem interacting with Database
+            DatasetNotFound: the specified dataset doesn't exist
+            DatasetTerminalStateViolation: dataset is in terminal state and
+                can't be advanced
+            DatasetBadStateTransition: dataset cannot be advanced to the
+                specified state
+
+        Returns:
+            Dataset: a dataset object in the desired state (if specified)
+        """
+        # Make sure we have controller and name from path
+        Dataset._render_path(kwargs)
+        controller = kwargs.get("controller")
+        name = kwargs.get("name")
+        try:
+            dataset = (
+                Database.db_session.query(Dataset)
+                .filter_by(controller=controller, name=name)
+                .first()
+            )
+        except SQLAlchemyError as e:
+            raise DatasetSqlError(f"Error attaching {controller}:{name}: {e}")
+        else:
+            if dataset is None:
+                raise DatasetNotFound(f"No dataset {controller}:{name}")
+            elif "state" in kwargs:
+                dataset.advance(kwargs.get("state"))
+        return dataset
+
+    @staticmethod
+    def attach_or_create(**kwargs):
+        """
+        attach_or_create Attempt to attach a dataset for the controller
+        and name. Note that `attach` is used to find an existing
+        dataset, and will attempt to advance it to the desired state if
+        found.
+
+        If the dataset does not currently exist in the state database, we'll
+        create a new one with the specified state value.
+
+        NOTE: In "steady state" we should always find a dataset tracker in
+        the database, and failure is a serious Pbench server error; TBD how
+        we want to handle that.
+
+        TODO: Unit tests often run individual server commands out of sequence
+        so that datasets that "should" exist don't exist; we often try to be
```

Comment:
Seems like we should tackle the unit tests to reign this in a bit.

Reply:
The problem is that it means inventing a new framework to add proper initial DB state to each test case, which I initially didn't consider and then found too imposing to tackle in the middle of this... although perhaps I should have gone that way first.

Reply:
So, OK; new "Simpled pickled Dataset DB state" commit includes `unittest` framework changes to support pre-loading a DB state using `pbench-state-manager`. This also now uses `attach` everywhere except prep-shim and PUT, which use `create`. (I haven't actually removed `attach_or_create` yet, but I think I can.) I haven't gotten through all of the unit tests yet, but interested parties can observe the progress.